BDA696 Final/final.py

```diff
@@ -0,0 +1,1245 @@
+    #
+    #     # Plot
+    #     trace_BF_cat = go.Heatmap(
+    #         z=sArr_cat_cat,
+    #         type="heatmap",
+    #     )
+    #     fig_BF_cat = go.Figure(data=[trace_BF_cat])
+    #     plot_BF_cat = f"BF-{tups_cat_cat[ind_cat_cat]}.html"
+    #     fig_BF_cat.write_html(
+    #         file=plot_BF_cat, include_plotlyjs="cdn"
+    #     )  # noqa
+    #     BF_plot["cat"] = plot_BF_cat
+    #     # fig_BF_cat.show()
+
+    # Machine Learning Models
+
+    # Defining the categorical columns
+    categoricalColumns = df_M.select_dtypes(include=[np.object]).columns
+
+    onehot_categorical = OneHotEncoder(handle_unknown="ignore")
+
+    categorical_transformer = Pipeline(steps=[("onehot", onehot_categorical)])
+
+    # Defining the numerical columns
+    numericalColumns = [
+        col
+        for col in df_M.select_dtypes(include=[np.float64]).columns
+        if col not in ["home_team_wins"]
+    ]
+
+    scaler_numerical = StandardScaler()
+
+    numerical_transformer = Pipeline(steps=[("scale", scaler_numerical)])
+
+    preprocessorForCategoricalColumns = ColumnTransformer(  # noqa
+        transformers=[("cat", categorical_transformer, categoricalColumns)],
+        remainder="passthrough",
+    )
+    preprocessorForAllColumns = ColumnTransformer(
+        transformers=[
+            ("cat", categorical_transformer, categoricalColumns),
+            ("num", numerical_transformer, numericalColumns),
+        ],
+        remainder="passthrough",
+    )
+
+    features = []
+    features = df_M.drop(["home_team_wins"], axis=1)
+
+    label = pd.DataFrame(df_M, columns=["price"])
+    label_encoder = LabelEncoder()
+    label = df_M["home_team_wins"]
+
+    label = label_encoder.fit_transform(label)
+
+    X_train, X_test, y_train, y_test = train_test_split(
+        features, label, random_state=0
+    )  # noqa
+
+    # K-Nearest classification model
+
+    model_name = "K-Nearest Neighbor Classifier"
+
+    knnClassifier = KNeighborsClassifier()
+
+    knn_model = Pipeline(
+        steps=[
+            ("preprocessorAll", preprocessorForAllColumns),
+            ("classifier", knnClassifier),
+        ]
+    )
+
+    knn_model.fit(X_train, y_train)
+
+    y_pred_knn = knn_model.predict(X_test)
+
+    y_test = label_encoder.transform(y_test)
+    y_test = label_encoder.inverse_transform(y_test)
+    y_pred_knn = label_encoder.inverse_transform(y_pred_knn)
+
+    Accuracy_score = round(accuracy_score(y_test, y_pred_knn), 2)
+    ax = plt.subplot()
+    fig = sns.heatmap(
+        confusion_matrix(y_test, y_pred_knn), annot=True, ax=ax, fmt=".1f"
+    ).get_figure()
+    ax.set_xlabel("Predicted labels")
+    ax.set_ylabel("True labels")
+    ax.set_title(
+        f"Confusion Matrix, {model_name}, Accuracy_score : {Accuracy_score}"
+    )  # noqa
+    ax.xaxis.set_ticklabels(["0", "1"])
+    ax.yaxis.set_ticklabels(["0", "1"])
+    fig.savefig(f"/results/Confusion Matrix-{model_name}.png")

```

Comment:
I like how these confusion matrix graphs came out. It would be nice if you could get rid of the redundant legends, but the format seems better than just the raw positive and negative predictive value method I used to check model performance.